pkg/operator/ceph/object/controller.go

```diff
@@ -283,18 +284,18 @@ func (r *ReconcileCephObjectStore) reconcile(request reconcile.Request) (reconci
 		// Remove finalizer
 		err = opcontroller.RemoveFinalizer(r.opManagerContext, r.client, cephObjectStore)
 		if err != nil {
-			return reconcile.Result{}, cephObjectStore, errors.Wrap(err, "failed to remove finalizer")
+			return reconcile.Result{}, *cephObjectStore, errors.Wrap(err, "failed to remove finalizer")
 		}
 
 		// Return and do not requeue. Successful deletion.
-		return reconcile.Result{}, cephObjectStore, nil
+		return reconcile.Result{}, *cephObjectStore, nil
 	}
 
 	if cephObjectStore.Spec.IsExternal() {
 		// Check the ceph version of the running monitors
 		desiredCephVersion, err := cephclient.LeastUptodateDaemonVersion(r.context, r.clusterInfo, config.MonType)
 		if err != nil {
-			return reconcile.Result{}, nil, errors.Wrapf(err, "failed to retrieve current ceph %q version", config.MonType)
+			return reconcile.Result{}, *cephObjectStore, errors.Wrapf(err, "failed to retrieve current ceph %q version", config.MonType)
```

Comment:
It looks like you added a check for nil in `ReportReconcileResult()`. Can't we just detect that case and skip the event report in that case, then we don't need to guarantee nil isn't passed?

Reply:
IMO, the nil check is not sufficient. It helps us not panic, but it doesn't actually enforce that we need an object to reference from an event. If the inner `reconcile()` doesn't return the object, we cannot report a reconcile succeeded/failed event for the reconcile, which means Rook is randomly dropping information we intend to deliver to the user.

Reply:
If the object is deleted, could it be sufficient to log the error? Or what's an example of why it's still critical to raise the event for a deleted object? If the log is sufficient, seems like we can keep this code simpler by skipping the event in this case.

Reply:
This isn't really about whether the object is deleted or not. It could be because the API server didn't return the object. In that case, we do still want to emit the ReconcileFailed event because it's important.

Reply:
Makes sense to protect against developer error, I was getting distracted by the rare deletion case.